### PR TITLE
Enhance Dockerfile to use multi-stage build

### PR DESCRIPTION
**Pull request by fpytloun**
This will build all assets and virtualenv and result in image that is
ready to be used for production use and has smallest possible size
(478MB).

I didn't tested docker-compose.yml changes much but I suppose it should be working for development in the same manner as before this change.

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [ ] Database queries are optimized and the number of queries is constant.
1. [ ] The changes are tested.
1. [ ] The code is documented (docstrings, project documentation).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,18 +1,56 @@
-FROM python:3.6
+### Build python virtualenv
+FROM python:3.6-stretch as build-python
 ENV PYTHONUNBUFFERED 1
 
 RUN \
   apt-get -y update && \
-  apt-get install -y gettext && \
-  apt-get clean
+  apt-get install -y gettext python3-virtualenv virtualenv && \
```
Comment (NyanKiyoshi) [inline Dockerfile]: Why installing both `python3-virtualenv` and `virtualenv`? `virtualenv` alone should be enough (which depends on 	`python3-virtualenv`).

File: Dockerfile [hunk 2]
```diff
@@ -1,18 +1,56 @@
...
+  apt-get clean && \
+  rm -rf /var/lib/apt/lists/*
```
Comment (NyanKiyoshi) [inline Dockerfile]: You could maybe add a comment saying `empty the APT cache`?

**Comment by patrys** [comment]
I'm not sure what to make of that. One one hand it provides a usable image after just one build. On the other hand:
* I'm not sure what we gain from introducing virtualenvs inside a container.
* We hardly ever deploy static assets as part of the docker container. They usually end up on S3 or a similar service and are served by the load balancer. It's a waste of resources to use a synchronous Python worker to serve all traffic.
* It looks like the entire virtualenv will be rebuilt each time Webpack needs to rebundle assets. This combined with the above makes me reluctant.

The build process becomes very complex and we don't seem to gain much. In the past we've discussed adding a separate docker-compose container to run `webpack --watch`.

**Comment by patrys** [comment]
@fpytloun Thank you for your first contribution to Saleor, please see above for some things I'd like to discuss before we decide how to proceed with your changes.

File: Dockerfile (hunk 1 above)
Comment (fpytloun) [reply Dockerfile]: This was required in jessie, you are right that it should not be needed now

File: Dockerfile (hunk 2 above)
Comment (fpytloun) [reply Dockerfile]: Ok

**Comment by patrys** [comment]
@fpytloun Do you have an example of what dev tooling is skipped when using virtualenv? I'm not sure I follow. All of our tooling is installed from `requirements.txt` so it would also end up in the virtualenv?

**Comment by fpytloun** [comment]
Image that's used to build virtualenv is python:3.6-stretch vs. final image is build using python:3.6-slim-stretch.
The first one contains build tools (pip, dev libs, etc.) so it is bigger. Second one doesn't have these tools but ships only bare interpreter. Check the size difference:
```
python                         3.6-slim-stretch    29ea9c0b39c6        7 weeks ago         138MB
python                         3.6-stretch         29d2f3226daf        7 weeks ago         912MB
```

So my image is 138MB + virtualenv size + assets size = only 478MB
When using `3.6-slim-stretch` and avoid using virtualenv, it would be over 1GB.

**Pull request merged**